### PR TITLE
bootstrap helper

### PR DESCRIPTION
**Pull request by Benjile**

### CODE REVIEW

File: tf [hunk 1]
```diff
@@ -75,7 +81,77 @@ function _tf_generic () {
...
+    echo "./.gitignore already present, skipping"
+  fi
+
+  # env directory
+  mkdir "${CONFIGURATION}" &>/dev/null || true
+  (
+    cd "${CONFIGURATION}"
+    if ! [[ -f "./shell.nix" ]]; then
+      # shell.nix
+      toolbox make-shell terraform tf kswitch
```
Comment (duthifree) [inline tf]: Problem with toolbox alias

File: tf [hunk 2]
```diff
@@ -75,7 +81,77 @@ function _tf_generic () {
...
+  else
+    echo "./envrc.EXAMPLE already present, skipping"
+  fi
+
+  # .gitignore
+  if ! [[ -f "./.gitignore" ]]; then
+    cat <<-'EOF' >"./.gitignore"
+			.terraform/
+			.envrc
+			.tmp/
```
Comment (eonpatapon) [inline tf]: Can you add `.direnv.d` also ?

File: tf [hunk 3]
```diff
@@ -75,7 +81,77 @@ function _tf_generic () {
...
+		EOF
+  else
+    echo "./.gitignore already present, skipping"
+  fi
+
+  # env directory
+  mkdir "${CONFIGURATION}" &>/dev/null || true
```
Comment (eonpatapon) [inline tf]: `mkdir -p ${CONFIGURATION}`

File: tf [hunk 4]
```diff
@@ -75,7 +81,77 @@ function _tf_generic () {
...
+    _tf_clone
+
+    # get envrc.EXAMPLE and tfvars file
+    for i in "./envrc.EXAMPLE" "./terraform.tfvars"; do
+      if ! [[ -f "${i}" ]]; then
+        cp "${TMP_DIR}/configurations/${CONFIGURATION}/${i}" . &>/dev/null || true
+      else
+        echo "./${CONFIGURATION}/${i} already present, skipping"
+      fi
+    done
```
Comment (eonpatapon) [inline tf]: I would do this the other way around:

```sh
CONFIG_DIR="${TMP_DIR}/configurations/${CONFIGURATION}"
for f in "envrc.EXAMPLE terraform.tfvars"; do
  if [ -f "${CONFIG_DIR}/${f}" ] && [ ! -f "${f}" ]; then
    cp "${CONFIG_DIR}/${f}" .
  fi
done
```

File: tf (hunk 1 above)
Comment (Benjile) [reply tf]: Seen with @eonpatapon 
toolbox should not been installed as an alias.

File: tf (hunk 2 above)
Comment (Benjile) [reply tf]: done

File: tf (hunk 3 above)
Comment (Benjile) [reply tf]: fixed

File: tf (hunk 4 above)
Comment (Benjile) [reply tf]: thanx, it is avoiding a false positive

File: tf [hunk 5]
```diff
@@ -75,7 +81,72 @@ function _tf_generic () {
...
+    cat <<-'EOF' >"./.envrc"
+			# creds for AMAZON S3 backend
+			# Those creds are individuals and should be stored in you personal keystore
+			# you can use gopass to retrieve them. For example:
+			export AWS_ACCESS_KEY_ID=$(gopass keystore/caascad/aws/181151069204/AWS_ACCESS_KEY_ID)
+			export AWS_SECRET_ACCESS_KEY=$(gopass keystore/caascad/aws/181151069204/AWS_SECRET_ACCESS_KEY)
+			
+			# creds for FLEXIBLE ENGINE provider
+			# Those creds are often available in the caascad keystore
+			# export TF_VAR_fe_access_key=$(gopass keystore/caascad/fe/OCB1111111/FE_ACCESS_KEY)
```
Comment (duthifree) [inline tf:96]: Today FE creds are not stored in personal keystore

File: tf [hunk 6]
```diff
@@ -75,7 +81,72 @@ function _tf_generic () {
...
+    cat <<-'EOF' >"./.gitignore"
+			.terraform/
+			.envrc
+			.tmp/
+			.direnv.d/
+		EOF
+  fi
+
+  # env directory
+  mkdir -p "${CONFIGURATION}" &>/dev/null || true
```
Comment (eonpatapon) [inline tf:112]:
```suggestion
  mkdir -p "${CONFIGURATION}"
```

Comment (eonpatapon) [reply tf:112]: mkdir -p never fails

File: tf [hunk 7]
```diff
@@ -75,7 +81,72 @@ function _tf_generic () {
...
+  (
+    cd "${CONFIGURATION}"
+    if ! [[ -f "./shell.nix" ]]; then
+      # shell.nix
+      toolbox make-shell terraform tf kswitch
+    fi
```
Comment (eonpatapon) [inline tf:118]: We don't really know which revision of the toolbox will work with the configuration, shouldn't shell.nix be provided by the configuration ?

File: tf [hunk 8]
```diff
@@ -75,7 +81,72 @@ function _tf_generic () {
...
+
+    # get the git repository
+    _tf_clone
+
+    # get envrc.EXAMPLE and tfvars file
+    CONFIG_DIR="${TMP_DIR}/configurations/${CONFIGURATION}"
+    for f in envrc.EXAMPLE terraform.tfvars.json terraform.tfvars; do
```
Comment (eonpatapon) [inline tf:125]: Copy shell.nix as well ?

File: tf (hunk 5 above)
Comment (Benjile) [reply tf]: fixed

File: tf (hunk 6 above)
Comment (Benjile) [reply tf]: done

File: tf (hunk 7 above)
Comment (Benjile) [reply tf:118]: yes you are absolutely right. I remove this

File: tf [hunk 9]
```diff
@@ -75,7 +81,68 @@ function _tf_generic () {
...
+  mkdir -p "${CONFIGURATION}"
+  (
+    cd "${CONFIGURATION}"
+
+    # get the git repository
+    _tf_clone
+
+    # get envrc.EXAMPLE and tfvars file
+    CONFIG_DIR="${TMP_DIR}/configurations/${CONFIGURATION}"
+    for f in envrc.EXAMPLE shell.nix *.tfvars*; do
```
Comment (eonpatapon) [inline tf:121]:
```suggestion
    for f in envrc.EXAMPLE shell.nix toolbox.json *.tfvars*; do
```
When creating a shell it also generates a `toolbox.json` file

File: tf [hunk 10]
```diff
@@ -75,7 +81,72 @@ function _tf_generic () {
...
+		EOF
+  fi
+
+  # env directory
+  mkdir -p "${CONFIGURATION}" &>/dev/null || true
+  (
+    cd "${CONFIGURATION}"
+    if ! [[ -f "./shell.nix" ]]; then
+      # shell.nix
+      toolbox make-shell terraform tf kswitch
```
[the remark this answers is not in the document]
Comment (Benjile) [reply tf]: i have removed the toolbox command due to jp comment. shell.nix should be provided by configurations

File: tf [hunk 11]
```diff
@@ -171,30 +248,38 @@ function _tf_parsing () {
...
         exit 1
       fi
+      ;;&
+    init | plan | apply | bootstrap)
       if [[ -z "${ENVIRONMENT}" ]]; then
         echo "Missing environment option"
         _tf_help
         exit 1
       fi
+      ;;&
```
[the remark this answers is not in the document]
Comment (Benjile) [reply tf:260]: from http://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Conditional-Constructs

> Using ‘;;&’ in place of ‘;;’ causes the shell to test the patterns in the next clause

File: tf (hunk 9 above)
Comment (Benjile) [reply tf]: done

**Pull request merged**